web/src/main/java/org/apache/shiro/web/util/WebUtils.java

```diff
@@ -92,6 +92,32 @@
      */
     public static final String DEFAULT_CHARACTER_ENCODING = "ISO-8859-1";
 
+    public static String getRemoteAddr(ServletRequest servletRequest) {
+        HttpServletRequest request = toHttp(servletRequest);
+        if (request == null) {
+            return "unknown";
+        }
+        String ip = request.getHeader("x-forwarded-for");
+        if (ip == null || ip.length() == 0 || "unknown".equalsIgnoreCase(ip)) {
+            ip = request.getHeader("Proxy-Client-IP");
+        }
+        if (ip == null || ip.length() == 0 || "unknown".equalsIgnoreCase(ip)) {
+            ip = request.getHeader("X-Forwarded-For");
```

Comment:
X-Forwarded-For could be a list:
https://en.wikipedia.org/wiki/X-Forwarded-For#Format